src/components/OpenLayersMap.js

```diff
@@ -12,154 +12,89 @@ import Projection from 'ol/proj';
-        }
-      });
-
-      _.each(drawFeatures, (f) => {
-        f.on('change', (e) => this.setState({modifying: e}));
-      });
-      modifyingSource.addFeatures(drawFeatures);
-    }
+  componentDidMount() {
+    this.setupMap();
   }
-  renderTrailsDebounce = _.debounce(this.renderTrails, 50);
 
   componentWillReceiveProps(nextProps) {
-    const {endDraw} = this.props;
-    const {map, interactions, modifyingSource} = this.state;
-    console.log(map);
-    // remove old draw interactions
-    interactions.forEach(interaction => {
-        map.removeInteraction(interaction);
-    });
+    const { interactions, source, map } = this.state;
+    const { canCreate, mode, createObject, trails, hydrants, selectedTrail } = nextProps;
+    // first sync to add externally-uploaded features or remove deleted ones
+    this.syncFeatures(trails, hydrants);
+
+    // remove all existing interactions
+    _.each(interactions, (interaction) => map.removeInteraction(interaction));
 
-    // creating new draw interactions
+    // create new draw or modify interaction depending on the props
     const newInteractions = [];
-    const newInteractionTypes = mapObjects[nextProps.mode] || [];
-    newInteractionTypes.forEach(type => {
+    if (canCreate) {
+      const type = mode === 'trails' ? 'Polygon' : 'Point';
       const draw = new Draw({
-        source: modifyingSource,
-        type: type,
-        geometryName: type
+        source, type, geometryName: type,
       });
-      draw.on('drawend', endDraw);
+      draw.on('drawend', (e) => createObject(e.feature));
       newInteractions.push(draw);
-    });
-    newInteractions.forEach(i => {
-      if (map) {
-        map.addInteraction(i);
+    }
+    if (selectedTrail) {
+      const modifiable = new Collection([]);
+      if (mode === 'trails') {
+        const feature = trails.getIn([selectedTrail, 'feature']);
+        modifiable.push(feature);
+      } else {
+        hydrants.filter((h) => h.get('trail') === selectedTrail)
+          .forEach((h) => modifiable.push(h.get('feature')));
       }
-    });
-    this.setState({interactions: newInteractions, modifying: null});
+      const modify = new Modify({ features: modifiable });
+      modify.on('modifyend', (e) => this.endModify(e));
+      newInteractions.push(modify);
+    }
 
-    this.renderTrailsDebounce(nextProps.trails, nextProps.selectedTrail);
+    _.each(newInteractions, (interaction) => map.addInteraction(interaction));
+    this.setState({ interactions: newInteractions });
   }
 
-  componentDidMount(){
-    this.setupMap();
+  componentDidUpdate(prevProps) {
+    const { selectedTrail, trails } = this.props;
+    const { map } = this.state;
+    if (selectedTrail !== prevProps.selectedTrail && selectedTrail) {
+      const firstCoords = trails.getIn([selectedTrail, 'coords'])[0];
+      const centerCoords = Projection.fromLonLat(firstCoords ? firstCoords : [0, 0]);
+      if (firstCoords) {
```

Comment:
What does animating the map view to coords [0,0] mean?

Reply:
I threw that in there as a workaround when it was failing for unrelated reasons in my `trail` objects. I'll remove it and make sure things work as they should again here.

Reply:
wurd.